kernel/sem.c

```diff
@@ -98,18 +113,165 @@ int z_vrfy_k_sem_init(struct k_sem *sem, unsigned int initial_count,
+			z_ready_thread(thread);
+			z_reschedule_no_lock(key);
+			return;
+		}
+		/* recheck count in case we lost the race condition
+		 * and all waiting threads timed out between us
+		 * incrementing count and actually checking the queue.
+		 */
+		count = atomic_get(&sem->count);
+	}
+
+	handle_poll_events(sem);
+	arch_irq_unlock(key);
+}
+
+static int sem_take(struct k_sem *sem, k_timeout_t timeout)
+{
+	int ret;
+	unsigned int key = arch_irq_lock();
+
+	if (K_TIMEOUT_EQ(timeout, K_NO_WAIT)) {
+		atomic_t count;
+		/* this doesn't truly need to be duplicated, but
+		 * we want as little as possible in the CASloop.
+		 */
+		do {
+			count = atomic_get(&sem->count);
+			if (count <= 0) {
+				ret = -EBUSY;
+				goto out;
+			}
+		} while (!atomic_cas(&sem->count, count, count - 1));
+		ret = 0;
+		goto out;
+	}
+
+	if (atomic_dec(&sem->count) > 0) {
+		ret = 0;
+		goto out;
+	}
+	ret = z_pend_curr_with_arch_lock(&sem->wait_q, timeout);
+
+	if (ret == -EAGAIN) {
+		/* We timed out, and now the semaphore value
+		 * is off by one because we're not in wait_q.
+		 * Give back our semaphore reservation.
+		 */
+		/* There is a lingering concern here with thread priorities.
+		 * If a bunch of low-priority threads do sem_take with a timeout,
+		 * and have enough cpu time to take the semaphore but not enough
+		 * to return it later, a high-priority thread may not be able
+		 * to take the semaphore until one of said low-priority threads
+		 * run. That being said, if you want priority inheritance you should
+		 * probably be using a mutex instead...
+		 * If this is an issue, we can fix it (at a performance penalty,
+		 * especially with the current APIs as there's no way to share the lock
+		 * call with the z_pend_curr call, and we'd likely be rescheduling twice
+		 * for much the same reason) by bumping this thread's priority for the
+		 * z_pend_curr call above.
+		 */
+		/*
+		 * Perhaps unintuatively, this
+		 * should _not_ do a normal semaphore give.
+		 */
+		atomic_inc(&sem->count);
+	}
+out:
+	arch_irq_unlock(key);
+	return ret;
+}
+
+static void sem_reset(struct k_sem *sem)
+{
+	atomic_t count;
+	atomic_t waiting;
+	bool reschedule_required = false;
+	unsigned int key = arch_irq_lock();
+
+	count = atomic_set(&sem->count, 0);
+
+	for (waiting = count; waiting < 0; waiting++) {
+		while (count < 0) {
+			struct k_thread *thread = z_unpend_first_thread(&sem->wait_q);
+			/* we can spin here for a moment if another
+			 * thread has updated count but
+			 * hasn't yet added itself to wait_q.
+			 */
+			if (thread != NULL) {
+				arch_thread_return_value_set(thread, -EAGAIN);
+				z_ready_thread(thread);
+				reschedule_required = true;
+				break;
+			}
+			/*
+			 * recheck count in case we lost the race condition
+			 * and all waiting threads timed out between us
+			 * incrementing count and actually checking the queue.
+			 */
+			count = atomic_get(&sem->count);
+		}
+	}
+	if (reschedule_required) {
```

Comment:
k_sem_reset() has not been traditionally documented as a scheduling point, and never acted as one.  I think you could skip this here, even though the newer code is indeed changing scheduler state.

Reply:
Previously `k_sem_reset` with pending threads was utterly broken and would silently cause all sorts of fun behavior, see #32899. (Actually, this PR subsumes the other PR, although I'm fine with either ordering of the two.)

This is essentially implementing `k_sem_reset` as an atomic version of

```c
while (!k_sem_get_count(sem)) {
    k_sem_give(sem);
}
while (k_sem_get_count(sem)) {
    k_sem_take(sem);
}
```
...which matches intuition here (for the most part). I could omit the reschedule, but that seems like a recipe for madness ("why does calling printf suddenly cause an unrelated thread to abort") and `k_sem_reset` doesn't seem like it's on the fastpath.

Reply:
It's more about implicit API promises.  It's less an issue on SMP where you always have to deal with races, but older Zephyr paradigms would lean heavily on cooperative thread scheduling, and the ability to know when you might be preempted.   Code written to that style would expect that its cleanup routine or whatever could do a k_sem_reset() without worrying about synchronization, now it needs to.

Is that a big deal in context?  Probably not.  But there's also no downside really.  Skipping a reschedule merely prevents whatever thread had been pended from running immediately on this CPU, it doesn't corrupt scheduler state or anything.  And on SMP it's largely a non-issue anyway as long as another CPU is available to take the IPI that was issued out of ready_thread().

It also removes lines, which is always nice.

Reply:
Fair enough.

- [ ] Remove reschedule from `k_spinlock_reset`
- [ ] Add comment to `k_spinlock_reset` about it not being a reschedule point.